src/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/XmlKeyHelper.cs

```diff
@@ -0,0 +1,319 @@
+                {
+                    // Make it so that if GetValue is called on
+                    // this name it'll advance into it correctly.
+                    _index--;
+                }
+
+                return ret;
+            }
+
+            internal string GetValue(string localName)
+            {
+                if (_enumerable == null)
+                {
+                    return null;
+                }
+
+                if (_enumerator == null)
+                {
+                    _enumerator = _enumerable.GetEnumerator();
+                }
+
+                int origIdx = _index;
+                int idx = origIdx;
+
+                if (!_enumerator.MoveNext())
+                {
+                    idx = -1;
+                    _enumerator = _enumerable.GetEnumerator();
+
+                    if (!_enumerator.MoveNext())
+                    {
+                        _enumerable = null;
+                        return null;
+                    }
+                }
+
+                idx++;
+
+                while (idx != origIdx)
+                {
+                    string curName = Functions.GetLocalName(_enumerator.Current);
+
+                    if (localName == curName)
+                    {
+                        _index = idx;
+                        return Functions.GetValue(_enumerator.Current);
+                    }
+
+                    if (!_enumerator.MoveNext())
+                    {
+                        idx = -1;
+
+                        if (origIdx < 0)
+                        {
+                            _enumerator = null;
+                            return null;
+                        }
+
+                        _enumerator = _enumerable.GetEnumerator();
+
+                        if (!_enumerator.MoveNext())
+                        {
+                            Debug.Fail("Original enumerator had elements, new one does not");
+                            _enumerable = null;
+                            return null;
+                        }
+                    }
+
+                    idx++;
+                }
+
+                return null;
+            }
+
+            private static class Functions
+            {
+                private static readonly Func<string, object> s_xDocumentCreate;
+                private static readonly PropertyInfo s_docRootProperty;
+                private static readonly MethodInfo s_getElementsMethod;
+                private static readonly PropertyInfo s_elementNameProperty;
+                private static readonly PropertyInfo s_nameNameProperty;
+                private static readonly PropertyInfo s_elementValueProperty;
+
+                static Functions()
+                {
+                    Type xDocument =
+                        Type.GetType(
+                            "System.Xml.Linq.XDocument, System.Private.Xml.Linq, Version=4.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51");
```

Comment:
can you depend on System.Xml.XDocument instead of this?

Reply:
No.  System.Private.Xml touches networking, networking touches crypto, so I can't touch System.Private.Xml.

Reply:
Is that only runtime dependency though? https://github.com/dotnet/corefx/blob/master/src/System.Xml.ReaderWriter/ref/System.Xml.ReaderWriter.csproj touches only System.Net.Primitives